compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -459,13 +457,184 @@ struct DistributeBroadcast final : OpDistributionPattern<vector::BroadcastOp> {
+    if (!srcLayout) {
+      return rewriter.notifyMatchFailure(multiReduceOp,
+                                         "expected nested layout attr");
+    }
+
+    VectorValue disSrc =
+        getDistributed(rewriter, srcVector, signature[srcVector]);
+    VectorValue disAcc =
+        getDistributed(rewriter, accVector, signature[accVector]);
+
+    Location loc = multiReduceOp.getLoc();
+
+    SmallVector<bool> reducedDims = multiReduceOp.getReductionMask();
+    int64_t rank = srcVector.getType().getRank();
+
+    // Do thread local reduce.
+
+    SmallVector<bool> distributedReductionMask;
+    distributedReductionMask.reserve(3 * rank);
+    distributedReductionMask.append(
+        applyPermutation(reducedDims, srcLayout.getBatchOrder()));
+    distributedReductionMask.append(
+        applyPermutation(reducedDims, srcLayout.getOuterOrder()));
+    distributedReductionMask.append(
+        applyPermutation(reducedDims, srcLayout.getElementOrder()));
+
+    auto localReduction = rewriter.create<vector::MultiDimReductionOp>(
+        loc, disSrc, disAcc, distributedReductionMask, multiReduceOp.getKind());
+    auto locallyReduced = dyn_cast<VectorValue>(localReduction.getResult());
+
+    if (!locallyReduced) {
+      return rewriter.notifyMatchFailure(
+          localReduction, "unimplemented: scalar result distribution");
+    }
+
+    // Flatten the locally reduced value.
+    VectorType shaped = locallyReduced.getType();
+    int64_t numElements = shaped.getNumElements();
+    SmallVector<int64_t> flatShape(1, numElements);
+    Type elemTy = shaped.getElementType();
+    VectorType flatVecType = VectorType::get(flatShape, elemTy);
+    VectorValue flat =
+        rewriter.create<vector::ShapeCastOp>(loc, flatVecType, locallyReduced);
+
+    FailureOr<VectorValue> globalReduced = doThreadGlobalReduction(
+        rewriter, srcLayout, flat, multiReduceOp.getKind(), reducedDims);
+    if (failed(globalReduced)) {
+      return failure();
+    }
+
+    VectorValue unflattened = rewriter.create<vector::ShapeCastOp>(
+        loc, shaped, globalReduced.value());
+    replaceOpWithDistributedValues(rewriter, multiReduceOp, unflattened);
+
+    return failure();
+  }
+
+  FailureOr<VectorValue>
+  doThreadGlobalReduction(RewriterBase &rewriter, NestedLayoutAttr layout,
+                          VectorValue flat, vector::CombiningKind kind,
+                          ArrayRef<bool> reductionMask) const {
+    VectorType flatVecType = flat.getType();
+    Type elemTy = flatVecType.getElementType();
+    int64_t numElements = flatVecType.getNumElements();
+    Location loc = flat.getLoc();
+
+    unsigned elemBitwidth = elemTy.getIntOrFloatBitWidth();
+    if (elemBitwidth != maxBitsPerShuffle) {
+      return rewriter.notifyMatchFailure(
+          flat.getDefiningOp(), llvm::formatv("unimplemented: packed shuffle",
+                                              elemBitwidth, maxBitsPerShuffle));
```

Comment:
Best practice is to not generate any IR before failing a pattern. Can you move this check to the beginning of the whole pattern?